### DISCUSSION TITLE
How to set up Amazon ElastiCache (Redis) Serverless

### DISCUSSION THREAD

**Discussion by mdobydullah** [laravel/framework#49501] [Q&A]
Hello Artisans,

I want to integrate [ElastiCache](https://aws.amazon.com/elasticache/) Serverless Redis. I've updated the Redis configurations and it works fine. 

I'm facing an issue with the Horizon queue:
```
Predis\Response\ServerException CROSSSLOT Keys in request don't hash to the same slot
/var/www/html/vendor/predis/predis/src/Client.php 370  
```

Sometimes, I'm experiencing these issues too:
```
Predis\Connection\ConnectionException Error while reading line from the server. 
[tcp://****.serverless.use2.cache.amazonaws.com:6379]
/var/www/html/vendor/predis/predis/src/Connection/AbstractConnection.php 155  
```

```
Illuminate\View\ViewException Error while reading line from the server.
[tcp://***.serverless.use2.cache.amazonaws.com:6379] (View:
/var/www/html/resources/views/frontend/errors/500.blade.php)
/var/www/html/vendor/predis/predis/src/Connection/AbstractConnection.php 155
```

### The config

config/database.php
```php
'redis' => [
    'client' => env('REDIS_CLIENT', 'predis'),

    'options' => [
        'cluster' => env('REDIS_CLUSTER', 'redis'),
        'prefix' => env('REDIS_PREFIX', Str::slug(env('APP_NAME', 'laravel'), '_').'_database_'),
    ],
    
    'clusters' => [
        'default' => [
            [
                'scheme' => env('REDIS_SCHEME', 'tcp'),
                'host' => env('REDIS_HOST', '127.0.0.1'),
                'password' => env('REDIS_PASSWORD', null),
                'port' => env('REDIS_PORT', 6379),
                'database' => 0,
            ],
        ],
        'cache' => [
            [
                'scheme' => env('REDIS_SCHEME', 'tcp'),
                'host' => env('REDIS_HOST', '127.0.0.1'),
                'password' => env('REDIS_PASSWORD', null),
                'port' => env('REDIS_PORT', 6379),
                'database' => 0,
            ],
        ],
    ],
],
```

.env
```
REDIS_SCHEME=tls
REDIS_HOST=*****.serverless.use2.cache.amazonaws.com
REDIS_PORT=6379
```

Please help me sort out the issues. Thanks in advance.

**Comment by yourdevDex** [comment]
Want to share, I tried this with scheme it connects when connect to redis cli with `--tls`. But still I encounter the error 

```
Error while reading line from the server.
[tcp://***.serverless.use2.cache.amazonaws.com:6379]
```

```
'default' => [
            'url' => env('REDIS_URL'),
            'host' => env('REDIS_HOST', '127.0.0.1'),
            'password' => env('REDIS_PASSWORD', null),
            'port' => env('REDIS_PORT', '6379'),
            'database' => env('REDIS_DB', '0'),
            'scheme' => 'tls',
        ],
```

**Comment by yourdevDex** [comment]
> Hello Artisans,
> 
> I want to integrate [ElastiCache](https://aws.amazon.com/elasticache/) Serverless Redis. I've updated the Redis configurations and it works fine.
> 
> I'm facing an issue with the Horizon queue:
> 
> ```
> Predis\Response\ServerException CROSSSLOT Keys in request don't hash to the same slot
> /var/www/html/vendor/predis/predis/src/Client.php 370  
> ```
> 
> Sometimes, I'm experiencing these issues too:
> 
> ```
> Predis\Connection\ConnectionException Error while reading line from the server. 
> [tcp://****.serverless.use2.cache.amazonaws.com:6379]
> /var/www/html/vendor/predis/predis/src/Connection/AbstractConnection.php 155  
> ```
> 
> ```
> Illuminate\View\ViewException Error while reading line from the server.
> [tcp://***.serverless.use2.cache.amazonaws.com:6379] (View:
> /var/www/html/resources/views/frontend/errors/500.blade.php)
> /var/www/html/vendor/predis/predis/src/Connection/AbstractConnection.php 155
> ```
> 
> ### The config
> config/database.php
> 
> ```
> 'redis' => [
>     'client' => env('REDIS_CLIENT', 'predis'),
> 
>     'options' => [
>         'cluster' => env('REDIS_CLUSTER', 'redis'),
>         'prefix' => env('REDIS_PREFIX', Str::slug(env('APP_NAME', 'laravel'), '_').'_database_'),
>     ],
>     
>     'clusters' => [
>         'default' => [
>             [
>                 'scheme' => env('REDIS_SCHEME', 'tcp'),
>                 'host' => env('REDIS_HOST', '127.0.0.1'),
>                 'password' => env('REDIS_PASSWORD', null),
>                 'port' => env('REDIS_PORT', 6379),
>                 'database' => 0,
>             ],
>         ],
>         'cache' => [
>             [
>                 'scheme' => env('REDIS_SCHEME', 'tcp'),
>                 'host' => env('REDIS_HOST', '127.0.0.1'),
>                 'password' => env('REDIS_PASSWORD', null),
>                 'port' => env('REDIS_PORT', 6379),
>                 'database' => 0,
>             ],
>         ],
>     ],
> ],
> ```
> 
> .env
> 
> ```
> REDIS_SCHEME=tls
> REDIS_HOST=*****.serverless.use2.cache.amazonaws.com
> REDIS_PORT=6379
> ```
> 
> Please help me sort out the issues. Thanks in advance.

Need to UP this or do you have already a solution?

**Reply by mdobydullah** [reply]
I've not found a solution yet.

**Reply by francirrenio** [reply]
Here's the translation of your message to English for posting as a comment on GitHub:

---

Here it's connecting normally using the method below:
```php
$this->redis = new Predis\Client("tls://url.serverless.use1.cache.amazonaws.com:6379");
```
In ElastiCache, it does not select the database. I'm using it without a password because the ElastiCache is in a private subnet and with access control by security group. If you use the `EXISTS` command, you will have to switch to `CF.EXISTS`, and the `KEYS` command for `SCAN`.

```php
private function getKeys($pattern){
    $timeout = 5;
    $startTime = microtime(true);
    $cursor = 0;
    $keys = [];
    do {
        $result = $this->redis->scan($cursor, 'match', $pattern, 'count', 100);
        $cursor = $result[0];
        $keys = array_merge($keys, $result[1]);

        // Checks if the timeout has been reached
        if (microtime(true) - $startTime > $timeout) {
            break;
        }
    } while ($cursor != 0);

    return $keys;
}
```

---

Feel free to adjust any technical terms if needed to better fit the context of your GitHub comment.

**Comment by 04116** [comment]
check this one https://github.com/phpredis/phpredis/issues/2083#issuecomment-1608711010

**Comment by evernest-com-au** [comment]
Adding tls:// to the front of the REDIS_HOST worked for me.  No need for the scheme in the config.

Eg:   Before: REDIS_HOST=host.serverless.use2.cache.amazonaws.com  after REDIS_HOST=tls://host.serverless.use2.cache.amazonaws.com

**Reply by MateusBMP** [reply]
It's worked for me with Laravel 11 and Redis OOS Caches engine 7.1 and solves the error "Error while reading line from the server". The error "CROSSSLOT Keys in request don't hash to the same slot" persists.

**Reply by markimpmc** [reply]
Did you manage to resolve the error "CROSSSLOT Keys in request don't hash to the same slot"?

I'm having the same error here trying to configure Laravel with Amazon Valkey.

**Comment by Orrison** [comment]
Does anyone else here have luck getting it all to work with Amazon ElastiCache (Redis) Serverless?

I have gotten it to work with the following configuration:
```
'redis' => [
        'client' => env('REDIS_CLIENT', 'phpredis'),

        'options' => [
            'cluster' => env('REDIS_CLUSTER', 'redis'),
            /** @phpstan-ignore-next-line */
            'prefix' => env('REDIS_PREFIX', Str::slug(env('APP_NAME', 'laravel'), '_') . '_database_'),
        ],

        'default' => [
            'scheme' => env('REDIS_SCHEME', 'tcp'),
            'host' => env('REDIS_HOST', '127.0.0.1'),
            'username' => env('REDIS_USERNAME'),
            'password' => env('REDIS_PASSWORD'),
            'port' => env('REDIS_PORT', '6379'),
            'database' => env('REDIS_DB', '0'),
        ],

        'cache' => [
            'scheme' => env('REDIS_SCHEME', 'tcp'),
            'url' => env('REDIS_URL'),
            'host' => env('REDIS_HOST', '127.0.0.1'),
            'username' => env('REDIS_USERNAME'),
            'password' => env('REDIS_PASSWORD'),
            'port' => env('REDIS_PORT', '6379'),
            'database' => env('REDIS_CACHE_DB', '1'),
        ],
    ],
```

I am using `phpredis` and have it working if a treat it like a single node.

It fails to connect and work if I treat it like a cluster by putting the `default` and `cache` items under `clusters` with an error `Error processing response from Redis node!`

So it has worked fine if I don't treat it like a cluster. But something doesn't feel right about that.
Occasionally (like 50 events in a second every couple of weeks) I will get an error `read error on connection to...` and `Redis::setex(): SSL: Connection reset by peer` but I am unsure if that is related to me not treating it like a cluster.

**Reply by sinemacula-ben** [reply]
@Orrison - I am encountering exactly the same issue!! Same setup (Elasticache) and occasional drops with the same error. I am about to start diving into this to try to determine what could be going wrong but figured I would ping here first - did you manage to identify the issue?

**Reply by Orrison** [reply]
I never did actually, ha! It plagues me and my error logs to this day. It doesn't happen very often and a lot of what we do with caching is okay if it misses, so it has never been prioritized.

If you figure it out that would be amazing!

**Reply by sinemacula-ben** [reply]
I'm on it - we use the cache a lot, and this is for a high-traffic production app, so it is a priority to us. Standby and I'll let you know what I find...

**Reply by sinemacula-ben** [reply]
I think I have solved it but I won't know for certain until it happens again (or doesn't happen) - the solution, if correct, is very simple... Basically, just utilize the built-in options available with phpredis and accessible through the Laravel config:

```
'redis' => [

    'client' => env('REDIS_CLIENT', 'phpredis'),

    'options' => [
        'cluster' => env('REDIS_CLUSTER', 'redis'),
        'prefix'  => env('REDIS_PREFIX', Str::slug(env('APP_NAME', 'laravel'), '_') . '_database_')
    ],

    'default' => [
        'url'               => env('REDIS_URL'),
        'host'              => env('REDIS_HOST', '127.0.0.1'),
        'username'          => env('REDIS_USERNAME'),
        'password'          => env('REDIS_PASSWORD'),
        'port'              => env('REDIS_PORT', '6379'),
        'database'          => env('REDIS_DB', '0'),
        'persistent'        => true,
        'timeout'           => env('REDIS_TIMEOUT', 1.0),
        'read_timeout'      => env('REDIS_READ_TIMEOUT', 1.0),
        'max_retries'       => env('REDIS_MAX_RETRIES', 5),
        'retry_interval'    => env('REDIS_RETRY_INTERVAL', 100),
        'backoff_algorithm' => env('REDIS_BACKOFF_ALGORITHM', 'decorrelated_jitter'),
        'backoff_base'      => env('REDIS_BACKOFF_BASE', 80),
        'backoff_cap'       => env('REDIS_BACKOFF_CAP', 1500),
        'tcp_keepalive'     => env('REDIS_TCP_KEEPALIVE', 60)
    ],

    'cache' => [
        'url'               => env('REDIS_URL'),
        'host'              => env('REDIS_HOST', '127.0.0.1'),
        'username'          => env('REDIS_USERNAME'),
        'password'          => env('REDIS_PASSWORD'),
        'port'              => env('REDIS_PORT', '6379'),
        'database'          => env('REDIS_CACHE_DB', '1'),
        'persistent'        => true,
        'timeout'           => env('REDIS_TIMEOUT', 1.0),
        'read_timeout'      => env('REDIS_READ_TIMEOUT', 1.0),
        'max_retries'       => env('REDIS_MAX_RETRIES', 5),
        'retry_interval'    => env('REDIS_RETRY_INTERVAL', 100),
        'backoff_algorithm' => env('REDIS_BACKOFF_ALGORITHM', 'decorrelated_jitter'),
        'backoff_base'      => env('REDIS_BACKOFF_BASE', 80),
        'backoff_cap'       => env('REDIS_BACKOFF_CAP', 1500),
        'tcp_keepalive'     => env('REDIS_TCP_KEEPALIVE', 60)
    ]

]
```

**Reply by Orrison** [reply]
Got it. I do remember messing around with a few of these, specifically `timeout` and `read_timeout`, hoping it would fix this issue based on some articles I had found. But that didn't seem to fix it at that time.

Are there any settings you are suggesting here that you feel is a most likely culprit for the issue?

Also are y'all customizing this beyond the defaults you have listed here with ENV values? (The setting specifically, of course y'all are most likely setting the URL, password, etc.)

**Reply by Orrison** [reply]
@sinemacula-ben did this issue ever come back for you?

Ad are you setting any of those env values to anything in particular that solves the issue?
